text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,234 @@
+└── lifecycle
+```
+### How will paths to dependencies be exposed to a buildpack?
+
+During the `build` phase of the `paketo-buildpacks/mri` buildpack, the `CNB_DEPS` environment variable will provide a path pointing to a `packaged.toml` file, this will contain information on all dependencies packaged with a buildpack.
+
+More on the contents of this file [here](#What-are-the-contents-of-the-packagedtoml-file-that-CNB_PATH-points-to).
+
+
+### How are dependencies added to a buildpackage, what is the `pack` user experience?
+These dependencies are added to a buildpackage by specifying fields in the `package.toml`
+
+``` toml
+[buildpack]
+uri = "./path/to/buildpack.tgz"
+
+[[dependencies]]
+uri = "path/to/dependency"
+id = "org.some-identifier"
+version = "v1.2.3"
+  [dependencies.metadata]
+  # dependency-specific data
+
+[[dependencies]]
+uri = "https://buildpacks.cloudfoundry.org/dependencies/node/node_10.20.0_linux_x64_cflinuxfs3_76f4a198.tgz"
+ID = "paketo.node"
+version = "v10.20.0"
+  [dependencies.metadata]
+  cool = "beans"
+
+
+```
+
+the above would give us the following file tree.
+
+```
+cnb
+└─ deps
+    ├── org.some-identifier
+    │    ├── v1.2.3.toml
+    │    └── v1.2.3
+    └── paketo.node
+         ├── v10.20.0.toml
+         └── v10.20.0
+
+```
+
+
+### How are dependencies grouped as layers
+Layers containing dependencies are attached as the final layers to a buildpackage. The exact distribution of dependencies into layers is left up to the packaging utility.
+
+
+### How are dependency layers matched with a buildpack?
+A buildpack packaged with dependencies will have the added `depLayers` field in the `org.buildpack.buildpack.layers` metadata.
+
+``` json 
+"paketo-buildpacks/node-engine":{
+      "0.0.178":{
+         "api":"0.2",
+         "stacks":[
+            {
+               "id":"org.cloudfoundry.stacks.cflinuxfs3"
+            },
+            {
+               "id":"io.buildpacks.stacks.bionic"
+            }
+         ],
+         "layerDiffID":"sha256:9ae53c280ea60b30b8c03978fbd2f761769fe0c15e421a9ece68e2f52926ce77",
+         "depLayers":{
+              "sha256:sha256-to-dep-layer": {
+                 "path": "/cnb/deps/path/to/dep1",
+                 "id": "org.some-identifier",
+                 "version": "v1.2.3", "metadata": {}
+              },
+              "sha256:sha256-to-other-dep-layer": {
+                 "path": "/cnb/deps/path/to/dep1",
+                 "id": "paketo.node",
+                 "version": "v10.20.0",
+                 "metadata": {"cool": "beans"}
+              }
+         }
+    
+      }
+```
+Each entry in the `depLayers` object maps a layer SHA to it's path. As well as preserves the `id`, `version` and `metadata` fields from `package.toml`
+
+### How do we determine if a dependency layer should be added to a build?
+The file tree rooted at the `/cnb/deps/` is only available at `build` time. So we add all layers to a build that match the following criteria:
```

Comment:
Does this have any implications for caching? Currently for `pack` we get a fair number of questions around e.g., why does the JDK get dowloaded separately for all of my apps... presumably if the dependency SHA is unchanged then that layer could get re-used across builds?

Reply:
There is now an alternative proposal that has behavior like this [here](https://github.com/dwillist/rfcs/blob/offline-buildpackages/text/0000-offline-buildpackages.md#always-require-offline_dependencies-would-need-a-name-change)